### ISSUE TITLE
Performance issues with useDragLayer hook

### ISSUE THREAD

**Issue by fromi** [react-dnd/react-dnd#2414]
**Describe the bug**
I am animating a hand of playing cards. I translate each card individually when it is dragged (instead of having a drag preview), therefore I have 1 custom drag layer per card.
The performances was dropping with the number of cards in hand, so I tried to avoid listening to the client offset in each drag layer. It turn out it is not possible given [the current implementation of useDragLayer hook](https://github.com/react-dnd/react-dnd/blob/master/packages/core/react-dnd/src/hooks/useDragLayer.ts#L10).

The 2 "useEffect" hooks inside useDragLayer looks pretty strange to me: why would you want to subscribe to something every time your component rerenders?
It turns out that inside the monitor, [we add a listener to the redux store each time](https://github.com/react-dnd/react-dnd/blob/master/packages/core/dnd-core/src/DragDropMonitorImpl.ts#L74).

I tried to reimplement useDragLayer but with only a single subscription when the hook is created: it does not work. I could not understand why the listener of the redux store is never called when I did that.

**Expected behavior**
I want to be able to collect the offset values conditionally in useDragLayer hook, in order to have multiple hooks but only one which causes the component to rerender.
Also, I think the spamming of listener into the redux store should be fixed to improve the performances.

**Workaround**
I created a custom specialized hook with performs much better:
```
const useDragOffsetDiff = (enabled: Boolean) => {
  const {dragDropManager} = useContext(DndContext)
  invariant(dragDropManager != null, 'Expected drag drop context')
  const monitor = (dragDropManager as DragDropManager).getMonitor()
  const [dragOffsetDiff, setDragOffsetDiff] = useState(monitor.getDifferenceFromInitialOffset())
  const offsetChangeListener = () => setDragOffsetDiff(monitor.getDifferenceFromInitialOffset())
  const [intervalId, setIntervalId] = useState<NodeJS.Timeout>()
  useEffect(() => {
    if (enabled) {
      setIntervalId(setInterval(offsetChangeListener, 40))
    } else if (intervalId) {
      clearInterval(intervalId)
    }
  }, [enabled])
  return dragOffsetDiff
}
```
Right now I could not find another way that setInterval to collect the offset. The store is not exposed by the DragDropManager, so I could not add my own listener to it.

**Comment by fromi** [comment]
I just came up with a better workaround, by listening to the redux store instead of setting an interval:
```
const useDragOffsetDiff = (enabled: Boolean) => {
  const {dragDropManager} = useContext(DndContext)
  invariant(dragDropManager != null, 'Expected drag drop context')
  const monitor = (dragDropManager as DragDropManager).getMonitor()
  const [dragOffsetDiff, setDragOffsetDiff] = useState(monitor.getDifferenceFromInitialOffset())
  const offsetChangeListener = () => {
    setDragOffsetDiff(monitor.getDifferenceFromInitialOffset())
  }
  const [unsubscribe, setUnsubscribe] = useState<Unsubscribe>()
  const store = ((monitor as any).store) as Store
  useEffect(() => {
    if (enabled) {
      const unsubscribe = store.subscribe(offsetChangeListener)
      setUnsubscribe(() => unsubscribe)
    } else if (unsubscribe) {
      unsubscribe()
    }
  }, [enabled])
  return dragOffsetDiff
}
```

**Comment by irrvrsl** [comment]
@fromi Can you share how to use your hook, please?

**Comment by fromi** [comment]
Sure:
```
  const dragOffsetDiff = useDragOffsetDiff(dragging)
```
And them you have dragOffsetDiff.x and dragOffsetDiff.y that you can use, for exemple to apply a CSS transform on the object that should follow the mouse: `transform: translate(${dragOffsetDiff?.x ?? 0}px, ${dragOffsetDiff?.y ?? 0}px);`

**Comment by hanh090** [comment]
> ```
>   const dragOffsetDiff = useDragOffsetDiff(dragging)
> ```
dragging value we get from hook `useDragLayer`? @fromi

**Comment by fromi** [comment]
@hanh090 Yes, using a collector as usual with React-DND:
```
const [{dragging}, ref] = useDrag({
    item,
    collect: monitor => ({
      dragging: monitor.isDragging()
    })
  })
```
Passing this value is important to prevent the interval to run forever when it is not useful.

**Comment by hanh090** [comment]
@fromi why you not use `monitor.isDragging` in your hook `useDragOffsetDiff`? besides, I see you use subscribe why it related to interval to run?

**Comment by dakir08** [comment]
Hi @fromi, could you please provide a full code with import as well?

**Comment by fromi** [comment]
Hi @dakir08, I updated my first comment to add the imports.
You can have a look at what I did with the drag and drop by going to [https://game-park.com/](https://game-park.com/) and playing the tutorial for It's a Wonderful World (it's a board game).

**Comment by dakir08** [comment]
@fromi Thanks so much :D. I can increase lots of performance with this :D

**Comment by jcampbell05** [comment]
I've done a tweak on the backend which seems to help see details for my comment https://github.com/react-dnd/react-dnd/issues/2621

**Comment by fromi** [comment]
I just came with a far better solution. This is how I use "useDragLayer" and prevent it from spamming too many changes:

```
import {shallowEqual} from '@react-dnd/shallowequal'
import {useState} from 'react'
import {DragLayerMonitor, useDragLayer} from 'react-dnd'

export default function useEfficientDragLayer<CollectedProps>(collect: (monitor: DragLayerMonitor) => CollectedProps): CollectedProps {
  const collected = useDragLayer(collect)
  const [previousCollected, setPreviousCollected] = useState<CollectedProps>(collected)
  const [requestID, setRequestID] = useState<number>()
  if (requestID === undefined && !shallowEqual(collected, previousCollected)) {
    setPreviousCollected(collected)
    setRequestID(requestAnimationFrame(() => setRequestID(undefined)))
  }
  return previousCollected
}
```

A few examples of drag&drop using this hook are available on my website: https://game-park.com/

**Comment by jcampbell05** [comment]
@fromi  nice any idea why that helps ?

**Comment by fromi** [comment]
> @fromi nice any idea why that helps ?

Yes, here is the general idea (I'm not a specialist): if I'm not mistaken, there is one process that runs the Javascript code, and one that refresh the screen.
The Javascript process runs a lot faster. With useDragLayer, it can spam new CSS values at a huge rate, which the display process cannot follow.
Using "requestAnimationFrame", you can slow it down to the rate of the display process.

**Comment by bipinrajbhar-rh** [comment]
> I just came with a far better solution. This is how I use "useDragLayer" and prevent it from spamming too many changes:
> 
> ```
> import {shallowEqual} from '@react-dnd/shallowequal'
> import {useState} from 'react'
> import {DragLayerMonitor, useDragLayer} from 'react-dnd'
> 
> export default function useEfficientDragLayer<CollectedProps>(collect: (monitor: DragLayerMonitor) => CollectedProps): CollectedProps {
>   const collected = useDragLayer(collect)
>   const [previousCollected, setPreviousCollected] = useState<CollectedProps>(collected)
>   const [requestID, setRequestID] = useState<number>()
>   if (requestID === undefined && !shallowEqual(collected, previousCollected)) {
>     setPreviousCollected(collected)
>     setRequestID(requestAnimationFrame(() => setRequestID(undefined)))
>   }
>   return previousCollected
> }
> ```
> 
> A few examples of drag&drop using this hook are available on my website: https://game-park.com/

If possible can you share a codesandbox example please?

**Comment by sKopheK** [comment]
@fromi unfortunately, your solution always returns the first item that was dragged, not the current one

**Comment by fromi** [comment]
@sKopheK maybe, anyway we moved to dndkit a while ago

**Comment by sKopheK** [comment]
thanks, i'll try that one
